BrainPortal/app/models/s3_data_provider.rb

```diff
@@ -40,137 +41,243 @@ def self.pretty_category_name #:nodoc:
-
-    Dir.chdir(Pathname.new(local_full).parent) do
-      if userfile.is_a?(FileCollection)
-        # Amazon does NOT provide chunked streaming.
-        # This means that IO.popen and File.popen both fail
-        # because they cannot provide a size for the content.
-        #src_fh = IO.popen("tar -czf - #{userfile.name.bash_escape}","r:BINARY")
-        # So, we have to make a local tarball instead. Hurgh.
-        system("tar", "-czf", tmp_tar_file, userfile.name)
-        src_fh = File.new(tmp_tar_file, "r:BINARY")
-      else
-        src_fh = File.new(local_full,"r:BINARY")
-      end
-      # bucket = @s3_connection.bucket.find(bucket_name)
-      @s3_connection.s3object.store(remote_filename, src_fh, bucket_name, :content_type => 'binary/octet-stream')
-      src_fh.close
+   
+    localfull      = cache_full_pathname(userfile)
+    remotefilename = provider_full_path(userfile)
+    
+    cb_error "Error: file #{localfull} does not exist in local cache" unless File.exists?(localfull)
+    if userfile.is_a?(FileCollection)
+      @s3_connection.copy_directory_to_bucket(localfull.to_s,File.dirname(remotefilename))
+    else
+      @s3_connection.copy_file_to_bucket(localfull.to_s,File.dirname(remotefilename))
     end
-    true
-  ensure
-    src_fh.close rescue true
-    File.unlink(tmp_tar_file) rescue true
-  end
+  end   
 
   def impl_provider_erase(userfile) #:nodoc:
     init_connection
-    @s3_connection.s3object.delete(s3_filename(userfile), bucket_name)
+    
+    remotefilename = provider_full_path(userfile)
+    @s3_connection.delete_path_from_bucket(remotefilename)
   end
 
   def impl_provider_rename(userfile,newname) #:nodoc:
     init_connection
-    @s3_connection.s3object.rename s3_filename(userfile), s3_filename(userfile,newname), bucket_name
-  end
+    
+    oldpath = provider_full_path(userfile)
+    remotedir = oldpath.parent
+    newpath = File.join(remotedir,newname)
 
-  def impl_provider_list_all(user) #:nodoc:
-    raise "Disabled"
-  #  init_connection
-  #  s3_connection.bucket.find(bucket_name).objects.map do |object|
-  #    file               = DataProvider::FileInfo.new()
-  #    filename           = filename_from_s3_filename(object.key)[1]
-  #    file.name          = filename
-  #    file.symbolic_type = :regular
-  #    file.mtime         = Time.parse(object.about()["date"]).to_i
-  #    file.size          = 0
-  #    file
-  #  end
+    if userfile.is_a?(FileCollection)
+      begin
+        @s3_connection.rename_path(oldpath,newpath)
+        return true
+      rescue
+        return false
+      end
+    else
+      begin
+        @s3_connection.rename_object(oldpath,newpath)
+        return true
+      rescue
+        return false
+      end
+    end
   end
 
-end
+  def impl_provider_list_all(user=nil) #:nodoc:
+
+    init_connection
+    fileData = @s3_connection.list_objects_short()
+    list = []
+    
+    # First parse the files
+    fileData[:files].each do |f|
```

Comment:
These two array code blocks are identical, except for the symbolic_type, can't we make them as one, and remove the redundancy?